### PR TITLE
Add buffer copy mode to only copy buffers from/to device when necessary and improve inference latency

### PR DESCRIPTION
**Pull request by YangleiZouIntel**
Only map the user buffers from&to host memory.

### CODE REVIEW

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (adstraw) [inline pmlc/rt/vulkan/vulkan_invocation.cc:795]: Is there a different way to handle this case?  Throwing here seems pretty severe.  Could we search from back-toward-front until we find a LaunchKernelAction?

**Review by adstraw — COMMENTED** [review]
Wondering about the `throw` but otherwise looks good.  Will leave final approval to @jbruestle based on the conversation about multiple-output nets, etc.

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.cc:795]: Yes, I agree we can search back to front and avoid throws exception. This patch is sort of a workaround I think we should check and map all buffers needed not only the last a few buffers as Jeremy mentioned in the meeting.

**Comment by adstraw** [comment]
Good ResNet50 results on my local machine (below).  Overall inference time, measured by stopwatch, down from about 1.5s to 1.1s with this change.

```
2020-09-15 11:08:40,953 VERBOSE-1 [default] Overall Vulkan time: 243.792ms
2020-09-15 11:08:40,953 VERBOSE-1 [default] Total Vulkan kernels: 1367
2020-09-15 11:08:40,953 VERBOSE-1 [default] Total Vulkan kernel exec time: 163.864ms
2020-09-15 11:08:40,953 VERBOSE-1 [default] Percentage Vulkan kernel exec time: 67.2148%
2020-09-15 11:08:40,953 VERBOSE-1 [default] Total Vulkan memory transfers: 1644
2020-09-15 11:08:40,953 VERBOSE-1 [default] Total (esimtated) Vulkan memory transfer time: 79.9276ms
2020-09-15 11:08:40,953 VERBOSE-1 [default] Percentage (estimated) Vulkan memory transfer time: 32.7852%
Example finished, elapsed: 12.281s (compile), 1.148s (execution)

-----------------------------------------------------------------------------------------
Network Name         Inference Latency         Time / FPS          
-----------------------------------------------------------------------------------------
resnet50             1148.32 ms                1148.32 ms / 0.87 fps
Correctness: PASS, max_error: 7.918380106275436e-06, max_abs_error: 1.4603137969970703e-06, fail_ratio: 0.0
```

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (jbruestle) [reply pmlc/rt/vulkan/vulkan_invocation.cc:795]: Yes.  We do know which buffers are user buffers (passed in as arguments) vs locals (alloced within the original function), so we should base the decision on that information rather than kernel #.  I'd rather not commit this change as as.

File: pmlc/rt/vulkan/wrappers.cc
Comment (adstraw) [inline pmlc/rt/vulkan/wrappers.cc:37]: Use BufferType instead of uint32_t

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:271]: wondering if we should use something like `isBlockArgument` and set to `true` here instead of the 1/0 uint32_t type system you have now.  Or, `isHostMemory` ?

**Review by adstraw — COMMENTED** [review]
Just a few small things about types and names.  I think it's OK to make this PR "ready for review"

File: pmlc/rt/vulkan/wrappers.cc
Comment (earhart) [reply pmlc/rt/vulkan/wrappers.cc:85]: Actually -- to keep the ABI really well-specified here, please do use uint32_t (but cast to `BufferType` immediately).

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.cc]: Thanks, right now map only user buffers. Resnet50 inference latency improved from about 650ms to 300ms. However, some keras backend tests failed,  checking now.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Yes, that is better. Using isBlockArgument and bool now.

File: pmlc/rt/vulkan/wrappers.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/wrappers.cc]: Updated with bool type isBlockArgument.

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.cc]: Some local buffers involved in calculations on the host function, I am working on adding logic to detect these buffers and copy them from/to device.

Example 1:
testSwitch: need to copy buffer %0 from host to device.
```
   func @get_value(%arg0: memref<2x3xf32> {tile.name = "anon"}, %arg1: memref<2x3xf32> {tile.name = "anon_0"}, %arg2: memref<f32> {tile.name = "anon_1"}, %arg3: memref<2x3xf32>) {
   %c1_i32 = constant 1 : i32
   %c0_i32 = constant 0 : i32
    %c1 = constant 1 : index
    %c2 = constant 2 : index
    %c3 = constant 3 : index
    %0 = alloc() : memref<i32>
    %1 = load %arg2[] : memref<f32>
    %2 = fptoui %1 : f32 to i1
    %3 = select %2, %c1_i32, %c0_i32 : i32
    store %3, %0[] : memref<i32>
    "gpu.launch_func"(%c1, %c1, %c1, %c2, %c3, %c1, %0, %arg1, %arg0, %arg3) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<i32>, memref<2x3xf32>, memref<2x3xf32>, memref<2x3xf32>) -> ()
    return
  }
```

Example 2:
testAvgPool: need to copy buffer %1 and %2 from device to host.
```
  func @get_value(%arg0: memref<1x4x4x1xf32> {tile.name = "anon"}, %arg1: memref<1x1x1x1xf32>) {
    %cst = constant 0.000000e+00 : f32
    %c0_i32 = constant 0 : i32
    %c0 = constant 0 : index
    %c4 = constant 4 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x4x4x1xi32>
    "gpu.launch_func"(%c1, %c1, %c1, %c4, %c4, %c1, %0) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>) -> ()
    "gpu.launch_func"(%c1, %c1, %c1, %c4, %c4, %c1, %0) {kernel = @get_value_kernel_0::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>) -> ()
    %1 = alloc() : memref<1x1x1x1xi32>
    store %c0_i32, %1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xi32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %0, %1) {kernel = @get_value_kernel_1::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xi32>, memref<1x1x1x1xi32>) -> ()
    %2 = alloc() : memref<1x1x1x1xf32>
    store %cst, %2[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %arg0, %2) {kernel = @get_value_kernel_2::@get_value_kernel} : (index, index, index, index, index, index, memref<1x4x4x1xf32>, memref<1x1x1x1xf32>) -> ()
    %3 = load %2[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    %4 = load %1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xi32>
    %5 = sitofp %4 : i32 to f32
    %6 = divf %3, %5 : f32
    store %6, %arg1[%c0, %c0, %c0, %c0] : memref<1x1x1x1xf32>
    return
  }
```

**Comment by adstraw** [comment]
@earhart does this latest commit work for you in terms of the ABI?

**Comment by earhart** [comment]
@adstraw No -- how wide is a bool?  I've used systems where it's a single byte, and where it's a full machine word, and where it could be either depending on the API (Windows NT, and yes, this caused bugs).

I think you should pick a fixed integer size for the ABI, and then use an enum internally (because bool parameters to APIs are also a source of bugs).

**Comment by earhart** [comment]
@adstraw @YangleiZouIntel If you do use bool, please make sure to intention comments on callees -- https://llvm.org/docs/CodingStandards.html#comment-formatting

**Comment by YangleiZouIntel** [comment]
> @adstraw @YangleiZouIntel If you do use bool, please make sure to intention comments on callers -- https://llvm.org/docs/CodingStandards.html#comment-formatting

Thanks @adstraw and @earhart. I checked the backend tests, it turned out we need at least 4 different buffer copy modes to get minimum number of buffer copies and achieve good inference latency. So I changed back to use uint32_t. Most backend tests are fixed now, still checking the last a few (<10) tests (which turned from pass to fail after this patch).

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.cc]: Thanks @jbruestle, the user buffers and the issues above are resolved.

**Comment by YangleiZouIntel** [comment]
Ready for review. Changing host buffer after shader submitted to device and before device returns is not allowed (generate warning)

File: pmlc/rt/vulkan/vulkan_invocation.h
Comment (earhart) [inline pmlc/rt/vulkan/vulkan_invocation.h:25]: Use an enum here.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (earhart) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:83]: I think these should be an enum attribute (generated by tblgen), but @jbruestle and @flaub might have other thoughts.  (And this could be done later if we decide to do it; just add a TODO.)

**Review by earhart — COMMENTED** [review]
Overall looks great!  Just a couple very minor comments.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:333]: You can `return copyMode` here as well.  Once we know there is a dependency outside the current block we are done.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:189]: It seems like there may be a design issue here if we need an (uninitialized) member variable to communicate between two member methods --- `convertGpuLaunchFunc` which sets `deinitVulkan` and `getBufferCopyMode` which consumes it.  I see that both methods are private and that ordering of calls to these methods is preserved by `runOnOperation` so it's not a bug, technically.  Again, just wondering if the need for this sort of thing indicates some design issue.  Could we reorganize code, somehow, to avoid this problem?

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:326]: I had some trouble reading this code given that the state we are conditioning on changes from line 324 (if dependency is before the *call op*) and line 326 (if *de initialization* is before the dependency).

I think it would read quite a lot more readable if you did this:
```
if dependency is before the call op then 
  set the copy-host-to-device bit
else // dependency is after the call op
  if de initialization is before the dependency then
    set the copy device-to-host bit
  else
    warning
```

I am wondering about this warning, though.  Under what circumstances would we hit this case?  Does it have to do with whether or not we are printing the buffers?  And, if it is just a warning is it worth checking?  Goes back to my design question from earlier --- maybe makes sense just to drop this check.  In which case you have...

```
if dependency is before the call op then 
  set the copy-host-to-device bit
else // dependency is after the call op
  set the copy device-to-host bit
```

And, no need to track the deinit across two member functions.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:266]: Just noticing that there is a lot of overlap between this function and the next function `isInternalOperation` and the analysis code in pmlc/dialect/pxa/analysis/uses.cc.  Wondering if we could / should find a more general place for this code so that it can be reused later.

**Review by adstraw — CHANGES_REQUESTED** [review]

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (earhart) [reply pmlc/rt/vulkan/vulkan_invocation.cc:26]: There's actually a tblgen thing to make it easy to generate enum attrs.  But IIRC there was some objection to defining new attributes; definitely check w/@flaub

Comment (flaub) [reply pmlc/rt/vulkan/vulkan_invocation.cc:26]: No objections from my end. I think we should use tblgen for generating enums as much as we can, since it provides lots of nice extra features (stringifying/symbolizing/documentation generation, etc)

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:83]: Fixed

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:322]: Fixed

File: pmlc/rt/vulkan/vulkan_invocation.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.cc:26]: Fixed. Thanks all.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:189]: Fixed and replaced "deinitVulkan" by "run", which is clearer.

File: pmlc/rt/vulkan/vulkan_invocation.h
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/vulkan_invocation.h]: Fixed

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Because currently we are doing calculations on both host and device side, thus some backend tests run into a situation where the host buffers are used after submit to vulkan and before returns, such as in testStd,
 
Below is a section of the IR for testStd
```
 func @get_value(%arg0: memref<3x4xf32> {tile.name = "anon"}, %arg1: memref<f32>) {
    %cst = constant 0.000000e+00 : f32
    %c12_i32 = constant 12 : i32
    %c0 = constant 0 : index
    %c3 = constant 3 : index
    %c4 = constant 4 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x1xf32>
    store %cst, %0[%c0, %c0] : memref<1x1xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %arg0, %0) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<1x1xf32>) -> ()
    %1 = alloc() : memref<1x1xf32>
    %2 = load %0[%c0, %c0] : memref<1x1xf32>
    %3 = sitofp %c12_i32 : i32 to f32
    %4 = divf %2, %3 : f32
    store %4, %1[%c0, %c0] : memref<1x1xf32>
    %5 = alloc() : memref<3x4xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c3, %c4, %arg0, %1, %5) {kernel = @get_value_kernel_0::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<1x1xf32>, memref<3x4xf32>) -> ()
    %6 = alloc() : memref<3x4xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c3, %c4, %5, %6) {kernel = @get_value_kernel_1::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<3x4xf32>) -> ()
    %7 = alloc() : memref<f32>
    store %cst, %7[] : memref<f32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %6, %7) {kernel = @get_value_kernel_2::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<f32>) -> ()
    %8 = alloc() : memref<f32>
    %9 = load %7[] : memref<f32>
    %10 = divf %9, %3 : f32
    store %10, %8[] : memref<f32>
    %11 = load %8[] : memref<f32>
    %12 = sqrt %11 : f32
    store %12, %arg1[] : memref<f32>
    return
  }
```
the code section below between the first and second gpu.launch_func is changing host side buffers after submit to vulkan, so it will end up with an incorrect result. But if the operation is stx.reshape, the result may still be correct.
```
    %1 = alloc() : memref<1x1xf32>
    %2 = load %0[%c0, %c0] : memref<1x1xf32>
    %3 = sitofp %c12_i32 : i32 to f32
    %4 = divf %2, %3 : f32
    store %4, %1[%c0, %c0] : memref<1x1xf32>
    %5 = alloc() : memref<3x4xf32>
```

I have detected a list of backend tests that currently has this issue, as follows,
FAIL: testBranchWithSoftmax (__main__.TestBackendOps) . 
FAIL: testRandomNormalDev (__main__.TestBackendOps) 
FAIL: testRandomUniformDev (__main__.TestBackendOps)  
FAIL: testRandomUniformVariableDev (__main__.TestBackendOps)  
FAIL: testStd (__main__.TestBackendOps) 
FAIL: testTruncatedNormalDev (__main__.TestBackendOps)  
FAIL: testVarSimple (__main__.TestBackendOps)  

Do you think put this message in IVLOG is better than warning?

To solve this issue we can 
1. Put everything between gpu.launch_func into a gpu module so they will be send to device
2. Rearrange the algorithm/pipeline passes to avoid changing host side buffers after copied to device
3. Make several vulkan command buffers and run the gpu.launch_funcs(and copy buffers back) before these host side ops.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: I think right now for debugging purposes, we have to check all dependencies to see if some of them changed host side buffers after copied to device. 
If we fixed this issue or we do not want to check here, we can early return copyMode.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (adstraw) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Addressing the larger design / testing issue in follow-on PRs works for me.  I see that you rewrote and commented the conditional logic in `getBufferCopyMode` and that looks good, thanks.

**Comment by adstraw** [comment]
Waiting for CI to pass before I do one final review but overall this looks good to me.  I would like if @earhart also reviewed and gave a 👍 to this PR when he's ready.

File: pmlc/conversion/gpu/enums.td
Comment (flaub) [inline pmlc/conversion/gpu/enums.td:8]: Use `CapitalCasing` instead of all caps.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:125]: Don't pass `Value` can `CallOp` by reference, always use by-value semantics.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:129]: Pass `Value` by-value

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:109]: Pass `CallOp` by value.

**Comment by flaub** [comment]
I feel like this entire functionality should be done at the `comp` dialect level.

**Comment by adstraw** [comment]
Changes LGTM 👍.  I have reset my review status so it no longer says "change requested".  Will leave it to either @flaub or @earhart to provide a second 👍 so this PR can land.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (earhart) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:200]: As long as you're changing this -- maybe use `llvm::StringSet`?  (It's not completely clear to me what the ownership rules are, here.)

File: pmlc/rt/vulkan/wrappers.cc
Comment (earhart) [inline pmlc/rt/vulkan/wrappers.cc:87]: Please make bufferCopyMode an int32_t (matching the call op), and then cast it before passing it to bindBuffer, so that the ABI is declared the same way on both the caller and callee.

**Review by earhart — COMMENTED** [review]
Overall looks good; just a few minor additional comments.

**Comment by YangleiZouIntel** [comment]
> I feel like this entire functionality should be done at the `comp` dialect level.

That makes a lot of sense. I am thinking about replace this pass with comp_to_vulkan pass.

File: pmlc/rt/vulkan/wrappers.cc
Comment (YangleiZouIntel) [reply pmlc/rt/vulkan/wrappers.cc]: Fixed, please check.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Fixed

File: pmlc/conversion/gpu/enums.td
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/enums.td]: Fixed

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Fixed

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:325]: Don't create pointers to `CallOp`, instead use it by-value.

Like:
```
LLVM::CallOp runOp = nullptr;
```

And later like:
```
runOp = llvmCallOp;
```

Comment (flaub) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:325]: Even better to extract this to a new function like:

```
Optional<LLVM::CallOp> findCallOp(llvm::StringRef name) {
}
```

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:456]: We should avoid using `MemRefCastOp`, the internal lowering of a MemRefCast is to use `alloca` memory (when lowered to LLVM). If this happens inside of a loop, it's easy for the stack to be exhausted.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc]: Fixed, added a findLLVMCallOp, since here only checks LLVM::CallOp

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:467]: I see, should we add a custom conversion pattern for MemRefCastOp to llvm and avoid use `alloca` memory?

Comment (YangleiZouIntel) [reply pmlc/conversion/gpu/gpu_to_vulkan.cc:467]: I see, this is a very good point, I will think more about it.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:324]: Need to return `llvm::None` here.

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:326]: Are you looking for the latest callOp or can you return early here?

File: pmlc/conversion/gpu/gpu_to_vulkan.cc
Comment (flaub) [inline pmlc/conversion/gpu/gpu_to_vulkan.cc:346]: Need to check the result here.

**Comment by flaub** [comment]
Is this PR still valid?

**Comment by YangleiZouIntel** [comment]
> Is this PR still valid?

No, this changes will be moved to comp dialect level. I will close this PR.

**Pull request closed without merging**
